File: Java/src/test/java/codingdojo/XMLExporterTest.java
package codingdojo;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.scrubbers.DateScrubber;
import org.approvaltests.scrubbers.RegExScrubber;

import java.util.Arrays;
import java.util.Collection;

import static codingdojo.SampleModelObjects.FlagshipStore;
import static codingdojo.SampleModelObjects.OldOrder;
import static codingdojo.SampleModelObjects.RecentOrder;

import org.junit.jupiter.api.Test;

public class XMLExporterTest {

    @Test
    public void exportFull() throws Exception {
        Collection<Order> orders = Arrays.asList(RecentOrder, OldOrder);
        String xml = XMLExporter.exportFull(orders);
        Approvals.verifyXml(xml);
    }

    @Test
    public void exportTaxDetails() throws Exception {
        Collection<Order> orders = Arrays.asList(RecentOrder, OldOrder);
        String xml = XMLExporter.exportTaxDetails(orders);
        Approvals.verifyXml(xml);
    }

    @Test
    public void exportStore() throws Exception {
        Store store = FlagshipStore;
        String xml = XMLExporter.exportStore(store);
        Approvals.verifyXml(xml);
    }

    @Test
    public void exportHistory() throws Exception {
        Collection<Order> orders = Arrays.asList(RecentOrder, OldOrder);
        String xml = XMLExporter.exportHistory(orders);
        String regex = "createdAt='[^']+'";
        Approvals.verifyXml(xml, new Options().withScrubber(
                DateScrubber.getScrubberFor("2022-06-21T11:29Z")
        ));
    }
}

File: Cpp/test/XMLExporterTest.cpp
#include "ApprovalTests.hpp"
#include "catch2/catch.hpp"

#include "SampleModelObjects.h"
#include "XMLExporter.h"

#include <vector>

std::string scrubDate(const std::string& input)
{
    static const std::regex regex("createdAt='[^']+'");

    int matchNumber = 1;
    auto result = input;
    std::smatch m;
    if (std::regex_search(result, m, regex))
    {
        auto guid_match = m.str();
        auto replacement = "createdAt='date_" + std::to_string(matchNumber) + "'";
        result = ApprovalTests::StringUtils::replaceAll(result, guid_match, replacement);
        matchNumber += 1;
    }
    return result;
}

void verifyXml(std::string xml, ApprovalTests::Options options=ApprovalTests::Options()) {
    //auto path = "/Users/emily/Library/Application\\ Support/JetBrains/Toolbox/apps/CLion/ch-0/201.7223.86/CLion.app";
    //auto args = "diff {Received} {Approved}";
    //auto reporter = ApprovalTests::CustomReporter::create(path, args);
    xml = ApprovalTests::StringUtils::replaceAll(xml, ">", ">\n");
    ApprovalTests::Approvals::verify(xml, options.fileOptions().withFileExtension(".xml"));
}



TEST_CASE("XMLExporter")
{
    make_sample_model_objects();

    SECTION("export store") {
        auto xml = XMLExporter().export_store(FlagshipStore);
        verifyXml(xml);
    }

   auto orders = std::vector<Order*>();
   orders.push_back(RecentOrder);
   orders.push_back(OldOrder);

   SECTION("export full") {
        auto xml = XMLExporter().export_full(orders);
        verifyXml(xml);
    }
    SECTION("export tax details") {
        auto xml = XMLExporter().export_tax_details(orders);
        verifyXml(xml);
   }
   SECTION("export history") {
        auto xml = XMLExporter().export_history(orders);
        auto options = ApprovalTests::Options().withScrubber(scrubDate);
        verifyXml(xml, options);
   }
}



